File: modules/geometry.lua
-- Pure geometry helpers used by the window-placement engine.
--
-- Everything in this file is a pure function over plain Lua tables:
--
--     frame = { x = number, y = number, w = number, h = number }
--     rect  = same shape
--
-- There are zero references to `hs.*`, so every function here is fully
-- testable under the Linux/busted runner.
--
-- Coordinate convention (matches macOS / Hammerspoon):
--
--   * Global coordinate space: origin (0, 0) is the top-left of the primary
--     screen. Secondary screens may have negative x or y.
--   * `frame` refers to a screen's *usable* area (excluding menu bar / Dock),
--     i.e. `hs.screen:frame()` on the real Mac.
--   * All rectangles are expressed in global coordinates unless explicitly
--     stated otherwise.

local M = {}

-- Epsilon used when rounding fractional pixels.
local EPS = 1e-6

local function round(n)
  if n >= 0 then
    return math.floor(n + 0.5 + EPS)
  else
    return -math.floor(-n + 0.5 + EPS)
  end
end

--- Make a new rectangle with rounded integer coordinates.
-- Hammerspoon accepts floats but macOS rounds internally; rounding here keeps
-- golden snapshots stable across platforms.
function M.rect(x, y, w, h)
  return { x = round(x), y = round(y), w = round(w), h = round(h) }
end

--- Shallow copy of a rect/frame.
function M.copy(r)
  return { x = r.x, y = r.y, w = r.w, h = r.h }
end

--- Structural equality for rectangles.
function M.equal(a, b)
  if a == nil or b == nil then return a == b end
  return a.x == b.x and a.y == b.y and a.w == b.w and a.h == b.h
end

--- Compute a sub-rectangle of `frame` from fractional coordinates.
-- `f = { x = x_frac, y = y_frac, w = w_frac, h = h_frac }` each in [0, 1].
-- Example: left half = { x = 0, y = 0, w = 0.5, h = 1 }.
function M.fractional(frame, f)
  return M.rect(
    frame.x + frame.w * f.x,
    frame.y + frame.h * f.y,
    frame.w * f.w,
    frame.h * f.h
  )
end

--- Cell in a uniform `cols x rows` grid on `frame`.
-- col, row are 1-indexed. span defaults to 1 cell.
function M.grid_cell(frame, cols, rows, col, row, span_cols, span_rows)
  assert(col >= 1 and col <= cols, "col out of range")
  assert(row >= 1 and row <= rows, "row out of range")
  span_cols = span_cols or 1
  span_rows = span_rows or 1
  return M.rect(
    frame.x + frame.w * (col - 1) / cols,
    frame.y + frame.h * (row - 1) / rows,
    frame.w * span_cols / cols,
    frame.h * span_rows / rows
  )
end

--- Halves: "left" / "right" / "top" / "bottom".
function M.half(frame, which)
  if which == "left" then
    return M.fractional(frame, { x = 0, y = 0, w = 0.5, h = 1 })
  elseif which == "right" then
    return M.fractional(frame, { x = 0.5, y = 0, w = 0.5, h = 1 })
  elseif which == "top" then
    return M.fractional(frame, { x = 0, y = 0, w = 1, h = 0.5 })
  elseif which == "bottom" then
    return M.fractional(frame, { x = 0, y = 0.5, w = 1, h = 0.5 })
  end
  error("unknown half: " .. tostring(which))
end

--- Quadrants: "nw" / "ne" / "sw" / "se".
function M.quadrant(frame, which)
  local mapping = {
    nw = { x = 0,   y = 0,   w = 0.5, h = 0.5 },
    ne = { x = 0.5, y = 0,   w = 0.5, h = 0.5 },
    sw = { x = 0,   y = 0.5, w = 0.5, h = 0.5 },
    se = { x = 0.5, y = 0.5, w = 0.5, h = 0.5 },
  }
  local f = mapping[which] or error("unknown quadrant: " .. tostring(which))
  return M.fractional(frame, f)
end

--- Horizontal thirds: "left" / "center" / "right".
function M.third(frame, which)
  local mapping = {
    left   = { x = 0,     y = 0, w = 1/3, h = 1 },
    center = { x = 1/3,   y = 0, w = 1/3, h = 1 },
    right  = { x = 2/3,   y = 0, w = 1/3, h = 1 },
  }
  local f = mapping[which] or error("unknown third: " .. tostring(which))
  return M.fractional(frame, f)
end

--- Two-thirds: "left" or "right".
function M.two_thirds(frame, which)
  if which == "left" then
    return M.fractional(frame, { x = 0, y = 0, w = 2/3, h = 1 })
  elseif which == "right" then
    return M.fractional(frame, { x = 1/3, y = 0, w = 2/3, h = 1 })
  end
  error("unknown two_thirds: " .. tostring(which))
end

--- Maximize within a screen's usable frame.
function M.maximize(frame)
  return M.copy(frame)
end

--- Keep current size, center on `frame`.
function M.centered(frame, rect)
  return M.rect(
    frame.x + (frame.w - rect.w) / 2,
    frame.y + (frame.h - rect.h) / 2,
    rect.w,
    rect.h
  )
end

--- Describe `rect` as fractions of `frame`. Inverse of `fractional`.
-- If `rect` extends beyond `frame`, fractions can exceed [0, 1]; that is
-- intentional so callers can choose to clamp or not.
function M.ratios_of(rect, frame)
  assert(frame.w > 0 and frame.h > 0, "frame must be non-empty")
  return {
    x = (rect.x - frame.x) / frame.w,
    y = (rect.y - frame.y) / frame.h,
    w = rect.w / frame.w,
    h = rect.h / frame.h,
  }
end

--- Clamp `rect` so it fits entirely within `frame`.
-- Width/height are preserved if possible, otherwise shrunk to fit.
function M.clamp(rect, frame)
  local w = math.min(rect.w, frame.w)
  local h = math.min(rect.h, frame.h)
  local x = math.max(frame.x, math.min(rect.x, frame.x + frame.w - w))
  local y = math.max(frame.y, math.min(rect.y, frame.y + frame.h - h))
  return M.rect(x, y, w, h)
end

--- Move `rect` to a different screen while preserving either
--   - `"ratios"`:  relative position & size on the target screen, or
--   - `"size"`:    exact pixel size, placed at the ratio-equivalent origin
--                  then clamped to fit, or
--   - `"center"`:  exact pixel size, centered on the target.
function M.move_to_screen(rect, from_frame, to_frame, mode)
  mode = mode or "ratios"
  if mode == "ratios" then
    return M.fractional(to_frame, M.ratios_of(rect, from_frame))
  elseif mode == "size" then
    local r = M.ratios_of(rect, from_frame)
    local target = M.rect(
      to_frame.x + to_frame.w * r.x,
      to_frame.y + to_frame.h * r.y,
      rect.w,
      rect.h
    )
    return M.clamp(target, to_frame)
  elseif mode == "center" then
    return M.centered(to_frame, rect)
  end
  error("unknown move_to_screen mode: " .. tostring(mode))
end

--- Nudge by one grid cell in a cardinal direction. `cols`/`rows` form a
-- virtual grid over `frame`. Movement is clamped to the frame.
function M.nudge(rect, frame, direction, cols, rows)
  cols = cols or 12
  rows = rows or 8
  local step_x = frame.w / cols
  local step_y = frame.h / rows
  local dx, dy = 0, 0
  if direction == "left"  then dx = -step_x
  elseif direction == "right" then dx = step_x
  elseif direction == "up"    then dy = -step_y
  elseif direction == "down"  then dy = step_y
  else error("unknown nudge direction: " .. tostring(direction)) end
  local target = M.rect(rect.x + dx, rect.y + dy, rect.w, rect.h)
  return M.clamp(target, frame)
end

--- Grow/shrink the rect width by one grid column, anchored on its left edge.
function M.resize_width(rect, frame, delta_cols, cols)
  cols = cols or 12
  local step = frame.w / cols
  local new_w = math.max(step, math.min(frame.w, rect.w + step * delta_cols))
  return M.clamp(M.rect(rect.x, rect.y, new_w, rect.h), frame)
end

--- Grow/shrink the rect height by one grid row, anchored on its top edge.
function M.resize_height(rect, frame, delta_rows, rows)
  rows = rows or 8
  local step = frame.h / rows
  local new_h = math.max(step, math.min(frame.h, rect.h + step * delta_rows))
  return M.clamp(M.rect(rect.x, rect.y, rect.w, new_h), frame)
end

--- Test if a rect is fully inside a frame (inclusive edges).
function M.contains(frame, rect)
  return rect.x >= frame.x
     and rect.y >= frame.y
     and rect.x + rect.w <= frame.x + frame.w
     and rect.y + rect.h <= frame.y + frame.h
end

--- Return the screen whose frame contains the largest portion of `rect`.
-- `screens` is a list of `{ frame = frame, ... }` items. Returns the index
-- (1-based) or nil if `screens` is empty.
function M.screen_for_rect(screens, rect)
  local best_idx, best_area = nil, -1
  for i, s in ipairs(screens) do
    local fx1, fy1 = s.frame.x, s.frame.y
    local fx2, fy2 = s.frame.x + s.frame.w, s.frame.y + s.frame.h
    local rx1, ry1 = rect.x, rect.y
    local rx2, ry2 = rect.x + rect.w, rect.y + rect.h
    local ox = math.max(0, math.min(fx2, rx2) - math.max(fx1, rx1))
    local oy = math.max(0, math.min(fy2, ry2) - math.max(fy1, ry1))
    local area = ox * oy
    if area > best_area then
      best_area = area
      best_idx = i
    end
  end
  return best_idx
end

return M

File: FreeWindow/FreeWindow/Core/Geometry.swift
// Geometry.swift — Pure geometry helpers for window placement.
// Direct 1:1 port of modules/geometry.lua.
// Zero dependencies on AppKit/Accessibility — fully testable.

import Foundation

public enum Half: String, CaseIterable {
    case left, right, top, bottom
}

public enum Quadrant: String, CaseIterable {
    case nw, ne, sw, se
}

public enum Third: String, CaseIterable {
    case left, center, right
}

public enum TwoThirdsSide: String, CaseIterable {
    case left, right
}

public enum Direction: String, CaseIterable {
    case left, right, up, down
}

public enum MoveMode: String {
    case ratios, size, center
}

public enum Geometry {

    // MARK: - Basic constructors

    /// Build a sub-rectangle of `frame` from fractional coordinates.
    public static func fractional(frame: WMRect, frac: FractionalRect) -> WMRect {
        WMRect(
            x: Double(frame.x) + Double(frame.w) * frac.x,
            y: Double(frame.y) + Double(frame.h) * frac.y,
            w: Double(frame.w) * frac.w,
            h: Double(frame.h) * frac.h
        )
    }

    /// Cell in a uniform cols×rows grid on `frame`. col, row are 1-indexed.
    public static func gridCell(frame: WMRect, cols: Int, rows: Int,
                         col: Int, row: Int,
                         spanCols: Int = 1, spanRows: Int = 1) -> WMRect {
        precondition(col >= 1 && col <= cols, "col out of range")
        precondition(row >= 1 && row <= rows, "row out of range")
        return WMRect(
            x: Double(frame.x) + Double(frame.w) * Double(col - 1) / Double(cols),
            y: Double(frame.y) + Double(frame.h) * Double(row - 1) / Double(rows),
            w: Double(frame.w) * Double(spanCols) / Double(cols),
            h: Double(frame.h) * Double(spanRows) / Double(rows)
        )
    }

    // MARK: - Halves

    public static func half(frame: WMRect, _ which: Half) -> WMRect {
        switch which {
        case .left:   return fractional(frame: frame, frac: FractionalRect(x: 0, y: 0, w: 0.5, h: 1))
        case .right:  return fractional(frame: frame, frac: FractionalRect(x: 0.5, y: 0, w: 0.5, h: 1))
        case .top:    return fractional(frame: frame, frac: FractionalRect(x: 0, y: 0, w: 1, h: 0.5))
        case .bottom: return fractional(frame: frame, frac: FractionalRect(x: 0, y: 0.5, w: 1, h: 0.5))
        }
    }

    // MARK: - Quadrants

    public static func quadrant(frame: WMRect, _ which: Quadrant) -> WMRect {
        let frac: FractionalRect
        switch which {
        case .nw: frac = FractionalRect(x: 0,   y: 0,   w: 0.5, h: 0.5)
        case .ne: frac = FractionalRect(x: 0.5, y: 0,   w: 0.5, h: 0.5)
        case .sw: frac = FractionalRect(x: 0,   y: 0.5, w: 0.5, h: 0.5)
        case .se: frac = FractionalRect(x: 0.5, y: 0.5, w: 0.5, h: 0.5)
        }
        return fractional(frame: frame, frac: frac)
    }

    // MARK: - Thirds

    public static func third(frame: WMRect, _ which: Third) -> WMRect {
        let frac: FractionalRect
        switch which {
        case .left:   frac = FractionalRect(x: 0,       y: 0, w: 1.0/3.0, h: 1)
        case .center: frac = FractionalRect(x: 1.0/3.0, y: 0, w: 1.0/3.0, h: 1)
        case .right:  frac = FractionalRect(x: 2.0/3.0, y: 0, w: 1.0/3.0, h: 1)
        }
        return fractional(frame: frame, frac: frac)
    }

    public static func twoThirds(frame: WMRect, _ which: TwoThirdsSide) -> WMRect {
        switch which {
        case .left:  return fractional(frame: frame, frac: FractionalRect(x: 0, y: 0, w: 2.0/3.0, h: 1))
        case .right: return fractional(frame: frame, frac: FractionalRect(x: 1.0/3.0, y: 0, w: 2.0/3.0, h: 1))
        }
    }

    // MARK: - Maximize & Center

    public static func maximize(frame: WMRect) -> WMRect {
        frame
    }

    public static func centered(frame: WMRect, rect: WMRect) -> WMRect {
        WMRect(
            x: Double(frame.x) + Double(frame.w - rect.w) / 2.0,
            y: Double(frame.y) + Double(frame.h - rect.h) / 2.0,
            w: Double(rect.w),
            h: Double(rect.h)
        )
    }

    // MARK: - Ratios

    /// Describe `rect` as fractions of `frame`. Inverse of `fractional`.
    public static func ratiosOf(rect: WMRect, frame: WMRect) -> FractionalRect {
        precondition(frame.w > 0 && frame.h > 0, "frame must be non-empty")
        return FractionalRect(
            x: Double(rect.x - frame.x) / Double(frame.w),
            y: Double(rect.y - frame.y) / Double(frame.h),
            w: Double(rect.w) / Double(frame.w),
            h: Double(rect.h) / Double(frame.h)
        )
    }

    // MARK: - Clamp

    /// Clamp `rect` so it fits entirely within `frame`.
    public static func clamp(rect: WMRect, frame: WMRect) -> WMRect {
        let w = min(rect.w, frame.w)
        let h = min(rect.h, frame.h)
        let x = max(frame.x, min(rect.x, frame.x + frame.w - w))
        let y = max(frame.y, min(rect.y, frame.y + frame.h - h))
        return WMRect(x: x, y: y, w: w, h: h)
    }

    // MARK: - Cross-screen move

    /// Move `rect` to a different screen while preserving relative position.
    public static func moveToScreen(rect: WMRect, from: WMRect, to: WMRect, mode: MoveMode = .ratios) -> WMRect {
        switch mode {
        case .ratios:
            return fractional(frame: to, frac: ratiosOf(rect: rect, frame: from))
        case .size:
            let r = ratiosOf(rect: rect, frame: from)
            let target = WMRect(
                x: Double(to.x) + Double(to.w) * r.x,
                y: Double(to.y) + Double(to.h) * r.y,
                w: Double(rect.w),
                h: Double(rect.h)
            )
            return clamp(rect: target, frame: to)
        case .center:
            return centered(frame: to, rect: rect)
        }
    }

    // MARK: - Nudge & Resize

    /// Nudge by one grid cell in a cardinal direction, clamped to frame.
    public static func nudge(rect: WMRect, frame: WMRect, direction: Direction, cols: Int = 12, rows: Int = 8) -> WMRect {
        let stepX = Double(frame.w) / Double(cols)
        let stepY = Double(frame.h) / Double(rows)
        var dx: Double = 0
        var dy: Double = 0
        switch direction {
        case .left:  dx = -stepX
        case .right: dx = stepX
        case .up:    dy = -stepY
        case .down:  dy = stepY
        }
        let target = WMRect(x: Double(rect.x) + dx, y: Double(rect.y) + dy, w: Double(rect.w), h: Double(rect.h))
        return clamp(rect: target, frame: frame)
    }

    /// Grow/shrink width by one grid column, anchored on left edge.
    public static func resizeWidth(rect: WMRect, frame: WMRect, deltaCols: Int, cols: Int = 12) -> WMRect {
        let step = Double(frame.w) / Double(cols)
        let newW = max(step, min(Double(frame.w), Double(rect.w) + step * Double(deltaCols)))
        let target = WMRect(x: Double(rect.x), y: Double(rect.y), w: newW, h: Double(rect.h))
        return clamp(rect: target, frame: frame)
    }

    /// Grow/shrink height by one grid row, anchored on top edge.
    public static func resizeHeight(rect: WMRect, frame: WMRect, deltaRows: Int, rows: Int = 8) -> WMRect {
        let step = Double(frame.h) / Double(rows)
        let newH = max(step, min(Double(frame.h), Double(rect.h) + step * Double(deltaRows)))
        let target = WMRect(x: Double(rect.x), y: Double(rect.y), w: Double(rect.w), h: newH)
        return clamp(rect: target, frame: frame)
    }

    // MARK: - Containment & screen detection

    /// Test if `rect` is fully inside `frame` (inclusive edges).
    public static func contains(frame: WMRect, rect: WMRect) -> Bool {
        rect.x >= frame.x &&
        rect.y >= frame.y &&
        rect.x + rect.w <= frame.x + frame.w &&
        rect.y + rect.h <= frame.y + frame.h
    }

    /// Return the screen index whose frame contains the largest overlap with `rect`.
    public static func screenForRect(screens: [ScreenInfo], rect: WMRect) -> Int? {
        var bestIdx: Int? = nil
        var bestArea: Int = -1
        for (i, s) in screens.enumerated() {
            let fx1 = s.frame.x
            let fy1 = s.frame.y
            let fx2 = s.frame.x + s.frame.w
            let fy2 = s.frame.y + s.frame.h
            let rx1 = rect.x
            let ry1 = rect.y
            let rx2 = rect.x + rect.w
            let ry2 = rect.y + rect.h
            let ox = max(0, min(fx2, rx2) - max(fx1, rx1))
            let oy = max(0, min(fy2, ry2) - max(fy1, ry1))
            let area = ox * oy
            if area > bestArea {
                bestArea = area
                bestIdx = i
            }
        }
        return bestIdx
    }
}
